pkg/controllermanager/controller/namespacedcloudprofile/reconciler.go

```diff
@@ -0,0 +1,161 @@
+	}
+
+	if err := patchCloudProfileStatus(ctx, r.Client, namespacedCloudProfile, parentCloudProfile); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// The deletionTimestamp labels the NamespacedCloudProfile as intended to get deleted. Before deletion, it has to be ensured that
+	// no Shoots and Seed are assigned to the NamespacedCloudProfile anymore. If this is the case then the controller will remove
+	// the finalizers from the NamespacedCloudProfile so that it can be garbage collected.
+	if namespacedCloudProfile.DeletionTimestamp != nil {
+		if !sets.New(namespacedCloudProfile.Finalizers...).Has(gardencorev1beta1.GardenerName) {
+			return reconcile.Result{}, nil
+		}
+
+		associatedShoots, err := controllerutils.DetermineShootsAssociatedTo(ctx, r.Client, namespacedCloudProfile)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if len(associatedShoots) == 0 {
+			log.Info("No Shoots are referencing the NamespacedCloudProfile, deletion accepted")
+
+			if controllerutil.ContainsFinalizer(namespacedCloudProfile, gardencorev1beta1.GardenerName) {
+				log.Info("Removing finalizer")
+				if err := controllerutils.RemoveFinalizers(ctx, r.Client, namespacedCloudProfile, gardencorev1beta1.GardenerName); err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+				}
+			}
+
+			return reconcile.Result{}, nil
+		}
+
+		message := fmt.Sprintf("Cannot delete NamespacedCloudProfile, because the following Shoots are still referencing it: %+v", associatedShoots)
+		r.Recorder.Event(namespacedCloudProfile, corev1.EventTypeNormal, v1beta1constants.EventResourceReferenced, message)
+		return reconcile.Result{}, fmt.Errorf("%s", message)
+	}
+
+	if !controllerutil.ContainsFinalizer(namespacedCloudProfile, gardencorev1beta1.GardenerName) {
+		log.Info("Adding finalizer")
+		if err := controllerutils.AddFinalizers(ctx, r.Client, namespacedCloudProfile, gardencorev1beta1.GardenerName); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func patchCloudProfileStatus(ctx context.Context, c client.Client, namespacedCloudProfile *gardencorev1beta1.NamespacedCloudProfile, parentCloudProfile *gardencorev1beta1.CloudProfile) error {
+	patch := client.StrategicMergeFrom(namespacedCloudProfile.DeepCopy())
+	MergeCloudProfiles(parentCloudProfile, namespacedCloudProfile)
+	namespacedCloudProfile.Status.CloudProfileSpec = parentCloudProfile.Spec
+	return c.Patch(ctx, namespacedCloudProfile, patch)
+}
+
+// MergeCloudProfiles builds the cloud profile spec from a base CloudProfile and a NamespacedCloudProfile by updating the CloudProfile Spec inplace.
+// The CloudProfile Spec can then be used as NamespacedCloudProfile.Status.CloudProfileSpec value.
+func MergeCloudProfiles(resultingCloudProfile *gardencorev1beta1.CloudProfile, namespacedCloudProfile *gardencorev1beta1.NamespacedCloudProfile) {
```

Comment:
```suggestion
func MergeCloudProfiles(namespacedCloudProfile *gardencorev1beta1.NamespacedCloudProfile, cloudProfile *gardencorev1beta1.CloudProfile) {
```

My impression is, that this signature would be clearer. Also, the function could directly merge into the status of the `NamespacedCloudProfile` instead of modifying the `CloudProfile` which seems a bit odd?!